torch/_refs/nn/functional/__init__.py

```diff
@@ -147,6 +149,64 @@ def elu(
     return torch.where(a > 0, a, rhs)
 
 
+# Pure Python implementation - don't register decomp
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("input", "target", "var"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def gaussian_nll_loss(
+    input: TensorLikeType,
+    target: TensorLikeType,
+    var: TensorLikeType,
+    full: bool = False,
+    eps: float = 1e-6,
+    reduction: str = "mean",
+) -> TensorLikeType:
+    # Check var shape
+    # var, which is a tensor of variances, can either represent a homoscedastic
+    # assumption (data has the same variance) or a heteroscedastic one (data has
+    # different variances).
+    #
+    # If var is the same shape as input, it's the heteroscedastic case.
+    # If var is *not* the same shape as input, it's the homoscedastic case.
+    #
+    # To support broadcasting, the following sub-cases are allowed in the
+    # homoscedastic case.  With all other sizes being the same as input (when
+    # comparing from the outermost dimension), var must either have:
+    # 1) one fewer dimension or
```

Comment:
More precisely, `input` and `var` represent the expectations and variances of a multivariate normal (AKA "Gaussian") distribution. The operation does NOT support traditional broadcasting between input and var. It does support

1. Input and var being the same shape
2. Input and var being the same shape except var has an innermost dimension of length one, which is then broadcast
3. Input and var being the same shape except var is missing the corresponding innermost dimension, in which case var is unsqueezed to create that innermost dimension and then it's broadcast as in (2)

If the variance is constant across distributions then a short-hand this allows to represent that is either:

1. `input` and `var` have the same shape, except `var` is missing the innermost dimension

Reply:
> Do input and var represent the expectations and variances of a single multivariate normal (AKA "Gaussian") distribution? Or are these the expectations and variances of multiple independent distributions, like one per each class?

I've looked at two arXiv papers linked from the original issue https://github.com/pytorch/pytorch/issues/48520 and the related PRs/issues, but I'm still not sure and would rather not change this part of the doc. I did remove the description from the functional op to have one source of truth.

Reply:
I understand but I think we should still do our best to describe the mathematics of what the operation does in this case -- that should lead to some interpretation?